python/google/protobuf/internal/reflection_test.py

```diff
@@ -1480,6 +1481,30 @@ def testMergeFromBug(self):
     message2.MergeFrom(message1)
     self.assertFalse(message2.HasField('optional_nested_message'))
 
+  def testReflectionDoesntBreakMerging(self):
+    """Test that building a class from reflection doesn't break merging."""
+    # Instantiate a TestMap from the generated class. (We use this class
+    # because copying a MessageMap relies on the map value's descriptor's
+    # concrete class, unlike most other copies, and is especially bug-prone)
+    proto = map_unittest_pb2.TestMap()
+    proto.map_string_nested_message["one"].c = 1234
+
+    print "First: %s" % proto
+
+    # Instantiate one from reflection
+    generated = reflection.MakeClass(proto.DESCRIPTOR)()
```

Comment:
The descriptor by design is only allowed to be bound with one concrete class e.g. we only have one concrete_class field. The generated message descriptor is already owned by the generated concrete class. As I mentioned in the previous comment, making a second class on the generated descriptor object is not supported. You would need to get the FileDescriptorProto out and build your own descriptor for the new class.

To really support making multiple classes on the same descriptor, we would have to redesign how descriptor and concrete class work. Otherwise it may simply break the python c++ extension based implementation, which relies the concrete class field. I don't think treating those two classes the same fixes the root problem.

I would suggest creating your own descriptors instead.

Reply:
Unfortunately, the Python protobuf code itself doesn't obey this invariant: as noted above, reflection.MakeClass() doesn't return an instance of the same class as is produced by the generated _pb2.py code. (They have the same class name but in different packages, and they share a descriptor)

The C++ implementation has a similar issue; there, reflection creates an object of some dedicated "reflected message" type, which is not the same as what shows up in the pb.cc's. We resolve it there the same way as we do here.

Reply:
I think then it's a documentation issue. reflection.MakeClass() requires the descriptors are built by the users, rather than from the protoc generated code. It should throw exceptions when the descritpor is already bound.

Reply:
Whoa, what? If that's true, then how are you supposed to parse a proto given its name? I'm doing the classic thing: storing the proto name and data type in an underlying store, then fetching the descriptor from the symbol database and passing it to reflection. 

To hand-build the descriptors would require copying a descriptor from the symbol database, which would require a deep-copy operation for descriptors (doesn't exist), and explicitly registering those copies in the symbol database (exists but is undocumented). And then you couldn't copy between a reflected and default object, which is especially weird: it's very natural to want to convert between the two, especially in a library.

The C++ code handles this very naturally, by treating the descriptor object as the proto identity; all these problems go away. (I'm painfully aware of this after having to factor out codex's crazy proto logic into SourceFileDatabase a few years ago) What I'm proposing is basically the identical behavior for the python code, and it makes everything "just work" the way callers expect it.

Reply:
I see. Sorry for the confusion. I thought you'd like to create your own classes. It seems like the reflection API was already broken by the concrete_class field. We need to either loosen the restriction by allowing multiple concrete classes (e.g. make that into a list), or introducing another API to retrieve the generated classes instead of creating new ones each time. I have forwarded this discussion to people working on protobuf python and they will chime in.

Reply:
Sorry for late reply. I looked into the code, does this solution sounds good for your usage? Add a check is the descriptor has '_concrete_class' already at the beginning of reflection.MakeClass():

  if hasattr(descriptor, '_concrete_class'):
    return descriptor._concrete_class

Your tests can pass with this solution.

Reply:
Done.

Reply:
This will change the behavior though, which has potential risks of breaking other users. 

We are still discussing about the proper fix for those APIs.  Mostly important question is about the semantics: shall we allow multiple classes created on the same descriptor. The implementation assumed this is disallowed, however we have two public APIs violating it, namely reflection.MakeClass and MessageFactory.GetPrototype. Also we need to converge the two APIs which basically are the same.

Reply:
/nod. That makes good sense. 

If it's any help, the C++ code didn't make this assumption as strongly as people had thought. The Python code primarily seems to do it through the _concrete_class variable, which isn't as widely depended on as it seems to be; I could imagine restructuring things to no longer depend on it.

But one invariant that turned out to make a lot of sense was that descriptor matching indicated proto _compatibility,_ for things like MergeFrom and so on, and was almost always more useful inside the proto code than class matching.

Reply:
Updates. We are still discuss the right way to fix it. There are actually two issues:
1, reflection.MakeClass() is duplicate with message_factory.GetPrototype(). I am contacting the author who added reflection.MakeClass() to know the history and see if this can be removed.
2, One of our main contributor has tried multiple times to kill _concrete_class without success. We are planning to give it another try.

Reply:
Well, I'd be happy to see _concrete_class go away, but that did look like a monstrous PITA. :)

FWIW, a fix very similar to this did work in C++, and has one important protection against breaking existing users: It causes something which previously didn't work (merging a reflected and constructed object) to work, but not vice-versa. So it would take a pretty exotic user (someone who was using a failure of MergeFrom to detect a reflected object?!) to be broken by this.

Reply:
We have marked the  reflection.MakeClass() deprecated and ask users to use MessageFactory.GetPrototype().

Is your test works well with MessageFactory.GetPrototype() without fix?